taichi/transforms/whole_kernel_cse.cpp

```diff
@@ -0,0 +1,67 @@
+#include "taichi/ir/ir.h"
+
+TLANG_NAMESPACE_BEGIN
+
+// Whole Kernel Common Subexpression Elimination
+class WholeKernelCSE : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+
+  WholeKernelCSE() {
+  }
+
+  void visit(IfStmt *if_stmt) override {
+    // Move common statements at the beginning or the end of both branches
+    // outside.
+    if (if_stmt->true_statements && if_stmt->false_statements) {
+      auto block = if_stmt->parent;
+      int current_stmt_id = block->locate(if_stmt);
+      auto &true_clause = if_stmt->true_statements;
+      auto &false_clause = if_stmt->false_statements;
+      if (irpass::same_statements(true_clause->statements[0].get(),
```

Comment:
`if_stmt->true_statements` might be an empty `std::vector`, then `true_clause->statements[0]` can lead to illegal memory access. Let's make sure the clauses are not empty. Same thing for `true_clause.back()`. (I triggered this while debugging `make_adjoint`.)